clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1611,4 +1611,154 @@ mlir::Value ScalarExprEmitter::VisitAbstractConditionalOperator(
 
         patchVoidOrThrowSites();
       });
-}
\ No newline at end of file
+}
+
+mlir::Value ScalarExprEmitter::VisitBinLAnd(const clang::BinaryOperator *E) {
+  if (E->getType()->isVectorType()) {
+    llvm_unreachable("NYI");
+  }
+
+  bool InstrumentRegions = CGF.CGM.getCodeGenOpts().hasProfileClangInstr();
+  mlir::Type ResTy = ConvertType(E->getType());
+  mlir::Location Loc = CGF.getLoc(E->getExprLoc());
+
+  // If we have 0 && RHS, see if we can elide RHS, if so, just return 0.
+  // If we have 1 && X, just emit X without inserting the control flow.
+  bool LHSCondVal;
+  if (CGF.ConstantFoldsToSimpleInteger(E->getLHS(), LHSCondVal)) {
+    if (LHSCondVal) { // If we have 1 && X, just emit X.
+
+      mlir::Value RHSCond = CGF.evaluateExprAsBool(E->getRHS());
+
+      if (InstrumentRegions) {
+        llvm_unreachable("NYI");
+      }
+      // ZExt result to int or bool.
+      return Builder.createZExtOrBitCast(RHSCond.getLoc(), RHSCond, ResTy);
+    }
+    // 0 && RHS: If it is safe, just elide the RHS, and return 0/false.
+    if (!CGF.ContainsLabel(E->getRHS()))
+      return Builder.getBool(false, Loc);
+  }
+
+  mlir::Value LHSCondV = CGF.evaluateExprAsBool(E->getLHS());
+  mlir::Value RHSCondV = CGF.evaluateExprAsBool(E->getRHS());
+  return Builder.create<mlir::cir::TernaryOp>(
+      Loc, LHSCondV, /*trueBuilder=*/
+      [&](mlir::OpBuilder &B, mlir::Location Loc) {
+        CIRGenFunction::LexicalScopeContext LexScope{Loc, Loc,
+                                                     B.getInsertionBlock()};
+        CIRGenFunction::LexicalScopeGuard lexElseGuard{CGF, &LexScope};
+        CGF.currLexScope->setAsTernary();
+        auto res = B.create<mlir::cir::TernaryOp>(
+            Loc, RHSCondV, /*trueBuilder*/
+            [&](mlir::OpBuilder &B, mlir::Location loc) {
+              CIRGenFunction::LexicalScopeContext lexScope{
+                  loc, loc, B.getInsertionBlock()};
```

Comment:
I'm not sure if this is the best way to do this. I also saw the comments in the ternary op. Do you want me to refactor out the get location method? If so, where would the best place to put the helper function?

Reply:
I was referring to using the same `loc` twice. I can copy the logic from the ternary op here if you prefer.

Reply:
> I was referring to using the same `loc` twice. 

Gotcha!

> I can copy the logic from the ternary op here if you prefer.

Yea, please copy it for now, that works.